### DISCUSSION TITLE
Are WebGPU supported ？

### DISCUSSION THREAD

**Discussion by catdoc** [RavEngine/RavEngine#35] [Q&A] [answered]
Are WebGPU supported ？
If supported, how to built it ?  Thank you for your help.

**Comment by Ravbug** [comment, marked as answer]
Hi, thank you for your interest in this project!

WebGPU support is an early work-in-progress and currently not functional. I can't provide an ETA for it because it's blocked by missing capabilities in WebGPU itself.
1. No bindless texture access. This is a showstopper because the engine's shadow system relies on it. There is an ongoing discussion regarding its addition to WebGPU: https://github.com/gpuweb/gpuweb/issues/380
2. No fragment shader interlock. This is also a showstopper because the current transparency system relies on it to work. There's an ongoing discussion regarding its addition to WebGPU: https://github.com/gpuweb/gpuweb/issues/395. For mobile devices the transparency system will require the use of Framebuffer fetch which is also not present in WebGPU: https://github.com/gpuweb/gpuweb/issues/396
3. No push constants. This one is an inconvenience that can be worked around. There is a proposal for adding it to WebGPU so if that lands then using that will be preferable to a custom emulation. https://github.com/gpuweb/gpuweb/blob/main/proposals/push-constants.md

I would like to finish WebGPU support for the engine, but I cannot until these issues with WebGPU are resolved.

**Reply by catdoc** [reply]
Thank you very much for your detailed response. This project is fantastic and an impressive piece of work!

Could RavEngine initially support the basic features of WebGPU? 
Since mainstream browsers already support WebGPU, would it be possible to implement alternative solutions for features like bindless textures and fragment shader interlock that you mentioned are lacking? 
Or could we consider temporarily not supporting those features in WebGPU? 

The basic functionality of WebGPU might be sufficient for most developers, at least for me. 
I look forward to RavEngine supporting WebGPU soon. Thanks again!

**Reply by Ravbug** [reply]
> would it be possible to implement alternative solutions for features like bindless textures and fragment shader interlock that you mentioned are lacking?

This would require implementing entirely different systems for shadows and transparency sorting, which represents a development and maintenance burden too significant for what is essentially a workaround to WebGPU deficiencies. 

> Or could we consider temporarily not supporting those features in WebGPU?

This is possible, but the result would be missing support for dynamic shadows and transparency in web builds, which I don't think would be very useful. If enough people want WebGPU support and are OK with this trade-off, I'll consider it.

**Comment by mpiccolino-tealblue** [comment]
Greetings, I'd like to test the engine for non-game simulations using WASM. For that kind of application, the transparency trade-off would be certainly ok. So +1! Keep up the excellent work.
